src/compute-plots.jl

```diff
@@ -0,0 +1,579 @@
+    register_colormapping!(attr, :image)
+    register_computation!(
+        attr,
+        [:x, :y],
+        [:data_limits],
+    ) do args, changed, _
+        mini_maxi = args[1][], args[2][]
+        mini = Vec3d(first.(mini_maxi)..., 0)
+        maxi = Vec3d(last.(mini_maxi)..., 0)
+        return (Rect3d(mini, maxi .- mini),)
+    end
+    T = typeof((attr[:x][], attr[:y][], attr[:image][]))
+    p = Plot{image,Tuple{T}}(user_kw, Observable(Pair{Symbol,Any}[]), Any[attr], Observable[])
+    p.transformation = Transformation()
+    return p
+end
+
+function compute_plot(::Type{Heatmap}, args::Tuple, user_kw::Dict{Symbol,Any})
+    attr = ComputeGraph()
+    add_attributes!(Heatmap, attr, user_kw)
+    register_arguments!(Heatmap, attr, user_kw, args...)
+    register_colormapping!(attr, :image)
+    register_computation!(attr, [:x, :y], [:data_limits]) do args, changed, _
+        mini_maxi = args[1][], args[2][]
+        mini = Vec3d(first.(mini_maxi)..., 0)
+        maxi = Vec3d(last.(mini_maxi)..., 0)
+        return (Rect3d(mini, maxi .- mini),)
+    end
+    T = typeof((attr[:x][], attr[:y][], attr[:image][]))
+    p = Plot{heatmap,Tuple{T}}(user_kw, Observable(Pair{Symbol,Any}[]), Any[attr], Observable[])
+    p.transformation = Transformation()
+    return p
+end
+
+function compute_plot(::Type{Scatter}, args::Tuple, user_kw::Dict{Symbol,Any})
+    attr = ComputeGraph()
+    add_attributes!(Scatter, attr, user_kw)
+    register_arguments!(Scatter, attr, user_kw, args...)
+    register_marker_computations!(attr)
+    register_colormapping!(attr)
+    register_computation!(attr, [:positions, :space, :markerspace, :quad_scale, :quad_offset, :rotation],
+                          [:data_limits]) do args, changed, last
+        return (_boundingbox(getindex.(args)...),)
+    end
+    T = typeof(attr[:positions][])
+    p = Plot{scatter,Tuple{T}}(user_kw, Observable(Pair{Symbol,Any}[]), Any[attr], Observable[])
+    p.transformation = Transformation()
+    return p
+end
+
+function attribute_per_pos!(attr, attribute::Symbol, output_name::Symbol)
+    register_computation!(
+        attr,
+        [attribute, :positions],
+        [output_name],
+    ) do (vec, positions), changed, last
+        vec[] isa AbstractVector || return (vec[],)
+        NP = length(positions[])
+        NC = length(vec[])
+        NP == NC && return (vec[],)
+        if NP ÷ 2 == NC
+            output = [vec[][div(i + 1, 2)] for i in 1:NP]
+            return (output,)
+        end
+        error("Color vector length $(NC) does not match position length $(NP)")
+        return (vec[],)
+    end
+end
+
+function compute_plot(::Type{Lines}, args::Tuple, user_kw::Dict{Symbol,Any})
+    attr = ComputeGraph()
+    add_attributes!(Lines, attr, user_kw)
+    register_arguments!(Lines, attr, user_kw, args...)
+    register_colormapping!(attr)
+    attribute_per_pos!(attr, :scaled_color, :synched_color)
+    attribute_per_pos!(attr, :linewidth, :synched_linewidth)
```

Comment:
This only applies to LineSegments. For them it maps a value per point pair to two values for the vertices, which results in per-segment colors/linewidths. In Lines this would result in a const value segment, then a gradient segment, then a const value segment, etc.